kernel/src/engine/arrow_expression.rs

```diff
@@ -215,17 +215,28 @@ fn ensure_data_types(kernel_type: &DataType, arrow_type: &ArrowDataType) -> Delt
             }
         }
         (DataType::Struct(kernel_fields), ArrowDataType::Struct(arrow_fields)) => {
-            require!(
-                kernel_fields.fields.len() == arrow_fields.len(),
-                make_arrow_error(format!(
-                    "Struct types have different numbers of fields. Expected {}, got {}",
-                    kernel_fields.fields.len(),
-                    arrow_fields.len()
-                ))
-            );
-            for (kernel_field, arrow_field) in kernel_fields.fields().zip(arrow_fields.iter()) {
+            // build a list of kernel fields that matches the order of the arrow fields
+            let mapped_fields = arrow_fields
+                .iter()
+                .flat_map(|f| kernel_fields.fields.get(f.name()));
+
+            // keep track of how many fields we matched up
+            let mut found_fields = 0;
+            // ensure that for the fields that we found, the types match
+            for (kernel_field, arrow_field) in mapped_fields.zip(arrow_fields) {
                 ensure_data_types(&kernel_field.data_type, arrow_field.data_type())?;
```

Comment:
I think we originally had this function returning the actual schema to use... but dropped that because it wasn't actually changing the schema. Ironically we may need to reinstate that control flow? That way, unexpected struct fields just get ignored because they're not in the returned schema?

(but agree that can be a follow-up after this PR fixes the immediate DAT issue)

Reply:
Yep, exactly. I started writing that for this PR and realized that munging arrow structs is... complicated, so decided to just do this quick fix.

We will need to make a deeper PR that will have to pass the actual `<dyn Array>` rather than just the schema, and then if the struct is having only a subset of fields selected, deconstruct each element, and put it back together without the dropped fields, and then return a new array to return as the result.

Reply:
Wait a minute... shouldn't we be filtering the read schema so it never even gets fetched from disk? Why would we need to filter arrow data at all?

Reply:
Great point! The issue is because we don't push down selection into leaf fields, and only support selection on the top level fields. Probably the _right_ thing to do is properly push that down, and then we don't  have to filter the arrow data, as you note, and this can probably get more simple again...